### PR TITLE
Add Prioritized DQN

### PR DESCRIPTION
**Pull request by tanzeyy**

### CODE REVIEW

File: examples/Prioritized_DQN/test_per.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/test_per.py]: All the test files' name must be end with `_test.py`

File: examples/Prioritized_DQN/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+## Prioritized Experience Replay
+Reproducing paper [Prioritized Experience Replay](https://arxiv.org/abs/1511.05952).
+
+Prioritized experience replay (PER) develops a framework for prioritizing experience, so as to replay important transitions more frequently. In original paper, the authors define the probability of sampling transition *i* as:
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/README.md]: Generally, in README.md, we show the link to the paper and describe the performance of algorithm. Please rewrite this file. Here is an [example](https://github.com/PaddlePaddle/PARL/tree/9444470c1f92f653e5e0c0d19c3bf44bc1cea665/examples/DQN_variant).

File: examples/Prioritized_DQN/atari/agent.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        pred_Q = np.squeeze(pred_Q, axis=0)
+        act = np.argmax(pred_Q)
+        return act
+
+    def learn(self, obs, act, reward, next_obs, terminal, ISweight):
+        if self.global_step % self.update_target_steps == 0:
+            self.alg.sync_target()
+        self.global_step += 1
+
+        lr = 0.00025 / 4
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/atari/agent.py]: When using a stable learning rate, we should put the definition of learning rate in algorithm. Please refer to the lr in PolicyGradient.
https://github.com/PaddlePaddle/PARL/blob/a50793e43acdbf5c27d5fd96bcec8b76dc777c0d/parl/algorithms/fluid/policy_gradient.py#L26

File: examples/Prioritized_DQN/atari/alg.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                -1,
+            ])
+        max_v.stop_gradient = True
+
+        target = reward + (
+            1.0 - layers.cast(terminal, dtype='float32')) * self.gamma * max_v
+        delta = layers.abs(target - pred_action_value)
+        cost = ISweight * layers.square_error_cost(pred_action_value, target)
+        cost = layers.reduce_mean(cost)
+        optimizer = fluid.optimizer.SGD(learning_rate=learning_rate)
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/atari/alg.py]: Why do we use SGD here? Adam is a faster optimizer in general.

File: examples/Prioritized_DQN/atari/agent.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        obs = np.expand_dims(obs, axis=0)
+        pred_Q = self.fluid_executor.run(
+            self.pred_program,
+            feed={'obs': obs.astype('float32')},
+            fetch_list=[self.value])[0]
+        pred_Q = np.squeeze(pred_Q, axis=0)
+        act = np.argmax(pred_Q)
+        return act
+
+    def learn(self, obs, act, reward, next_obs, terminal, ISweight):
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/atari/agent.py]: Please name the arguments using the same style. IsWeight -> sample_weight.

File: examples/Prioritized_DQN/atari/alg.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+import numpy as np
+import paddle.fluid as fluid
+
+import parl
+from parl.core.fluid import layers
+from parl.core.fluid.algorithm import Algorithm
+
+
+class PrioritizedDQN(parl.Algorithm):
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/atari/alg.py]: Please put the file to the algorithm folder so that users can import it from parl.algorithm.
folder: parl/algorithms/fluid

File: examples/Prioritized_DQN/train.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        '--train_total_steps',
+        type=int,
+        default=int(1e7),
+        help='maximum environmental steps of games')
+    parser.add_argument(
+        '--test_every_steps',
+        type=int,
+        default=100000,
+        help='the step interval between two consecutive evaluations')
+    parser.add_argument(
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/train.py]: Please remove the argument.

File: examples/Prioritized_DQN/train.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        mem = []
+        while total_step < MEMORY_WARMUP_SIZE:
+            total_reward, steps, _ = run_episode(
+                env, agent, per, mem=mem, warmup=True)
+            total_step += steps
+            pbar.update(steps)
+    per.elements.from_list(mem[:int(MEMORY_WARMUP_SIZE)])
+
+    log_dir = './exps/{}/'.format(args.exp_name)
+    logger.set_dir(log_dir)
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/train.py]: just use `auto_set_dir()` here.

File: examples/Prioritized_DQN/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
++ atari-py
++ [ale_python_interface](https://github.com/mgbellemare/Arcade-Learning-Environment)
+
+
+### Start Training:
+Train on BattleZone game with DDQN and rank-based PER:
+```bash
+python train.py --rom ./rom_files/battle_zone.bin --alg ddqn --prior_type rank --exp_name run_train
+```
+training and evaluation curves will be saved in `./exps/run_train`, which can viewed with:
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/README.md]: Please remove the block as PARL doesn't support visualdl at every platform(only at linux with python3.6+).

File: examples/Prioritized_DQN/per/proportional.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        return clipped
+
+    def update(self, indices, priorities):
+        priorities = np.array(priorities) + self.eps
+        priorities_alpha = np.power(priorities, self.alpha)
+        for idx, priority in zip(indices, priorities_alpha):
+            self.elements.update(idx, priority)
+            self._max_priority = max(priority, self._max_priority)
+
+    @_check_full
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/per/proportional.py]: Please remove the decorator here. We are writing examples for the users. We have to reduce the bar to understand the code. A decorator can make the code cooler, but harder to understand.

**Review by TomorrowIsAnOtherDay — CHANGES_REQUESTED** [review]

File: examples/Prioritized_DQN/per/baselines/replay_memory.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
+# Third party code
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/per/baselines/replay_memory.py]: `baselines` directory is not used, please remove it.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

**Comment by tanzeyy** [comment]
Modified as required, experiment results on atari environment will be added when finished.

File: examples/Prioritized_DQN/README.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
+## Prioritized Experience Replay
+Reproducing paper [Prioritized Experience Replay](https://arxiv.org/abs/1511.05952).
+
+Prioritized experience replay (PER) develops a framework for prioritizing experience, so as to replay important transitions more frequently. In original paper, the authors define the probability of sampling transition *i* as:
+
+<img src=".figs/prob.jpg" />
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/README.md]: The size of these images may be too large, please reduce it.

File: examples/Prioritized_DQN/atari/agent.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        pred_Q = np.squeeze(pred_Q, axis=0)
+        act = np.argmax(pred_Q)
+        return act
+
+    def learn(self, obs, act, reward, next_obs, terminal, sample_weight):
+        if self.global_step % self.update_target_steps == 0:
+            self.alg.sync_target()
+        self.global_step += 1
+
+        lr = self.alg.lr
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/atari/agent.py]: It's redundant. Please don't feed the learning rate to the algorithm.

File: examples/Prioritized_DQN/atari/alg.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+# limitations under the License.
+
+import copy
+
+import numpy as np
+import paddle.fluid as fluid
+
+import parl
+from parl.core.fluid import layers
+from parl.core.fluid.algorithm import Algorithm
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/atari/alg.py]: Please remove this line, it's not used.

File: examples/Prioritized_DQN/atari/alg.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+    def learn(self, obs, action, reward, next_obs, terminal, learning_rate,
+              sample_weight):
+        pred_value = self.model.value(obs)
+        action_onehot = layers.one_hot(action, self.act_dim)
+        action_onehot = layers.cast(action_onehot, dtype='float32')
+        pred_action_value = layers.reduce_sum(
+            layers.elementwise_mul(action_onehot, pred_value), dim=1)
+
+        # choose acc. to behavior network
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/atari/alg.py]: The comment is not easy to understand. Please clarify it.

File: examples/Prioritized_DQN/atari/alg.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    def predict(self, obs):
+        return self.model.value(obs)
+
+    def learn(self, obs, action, reward, next_obs, terminal, learning_rate,
+              sample_weight):
+        pred_value = self.model.value(obs)
+        action_onehot = layers.one_hot(action, self.act_dim)
+        action_onehot = layers.cast(action_onehot, dtype='float32')
+        pred_action_value = layers.reduce_sum(
+            layers.elementwise_mul(action_onehot, pred_value), dim=1)
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/atari/alg.py]: Can we directly use `*` instead of `elementwise_mul`?

File: examples/Prioritized_DQN/per/utils.py [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+        self.alpha = alpha
+        self.seg_num = seg_num
+        self.size = int(size)
+        self.elements = elem_cls(self.size)
+        if init_mem:
+            self.elements.from_list(init_mem)
+        self.framestack = framestack
+        self._max_priority = 1.0
+
+    def _check_full(self):
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/per/utils.py]: The implementation is not compact, maybe can replace it with 
`assert self.elements.full(), "The replay memory is not full!"`

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: examples/Prioritized_DQN/proportional_per.py [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+
+class SumTree:
```
Comment (TomorrowIsAnOtherDay) [inline examples/Prioritized_DQN/proportional_per.py]: SumTree(object)

File: examples/Prioritized_DQN/README.md [hunk 19]
```diff
@@ -0,0 +1,30 @@
+## Prioritized Experience Replay
+Reproducing paper [Prioritized Experience Replay](https://arxiv.org/abs/1511.05952).
+
+Prioritized experience replay (PER) develops a framework for prioritizing experience, so as to replay important transitions more frequently. There are two variants of prioritizing the transitions, rank-based and proportional-based. Our implementation is the proportional variant, which has a better performance, as reported in the original paper.
+
+## Reproduced Results
+Results have been reproduced on following three environments:
```
Comment (zenghsh3) [inline examples/Prioritized_DQN/README.md]: please specify the algorithm variant is dqn or double dqn.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.

**Review by zenghsh3 — APPROVED** [review]
LGTM
